contracts-periphery/src/UmbraBatchSend.sol

```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.13;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+interface IUmbra {
+  function sendEth(
+    address payable receiver,
+    uint256 tollCommitment,
+    bytes32 pkx,
+    bytes32 ciphertext
+  ) external payable;
+
+  function sendToken(
+    address receiver,
+    address tokenAddr,
+    uint256 amount,
+    bytes32 pkx,
+    bytes32 ciphertext
+  ) external payable;
+}
+
+contract UmbraBatchSend {
+  IUmbra internal constant umbra = IUmbra(0xFb2dc580Eed955B528407b4d36FfaFe3da685401);
+
+  struct SendEth {
+    address payable receiver;
+    uint256 amount;
+    bytes32 pkx;
+    bytes32 ciphertext;
+  }
+
+  struct SendToken {
+    address receiver;
+    address tokenAddr;
+    uint256 amount;
+    bytes32 pkx;
+    bytes32 ciphertext;
+  }
+  
+  error ValueMismatch();
+  event Log(address indexed caller, uint256 indexed value, string message);
+
+  function batchSendEth(uint256 _tollCommitment, SendEth[] calldata _params) public payable {
+    uint256 valAccumulator;
+
+    for (uint256 i = 0; i < _params.length; i++) {
+      //amount to be sent per receiver
+      uint256 _amount = _params[i].amount;
+      valAccumulator += _amount;
+
+      umbra.sendEth{value: _amount}(_params[i].receiver, _tollCommitment, _params[i].pkx, _params[i].ciphertext);
+    }
+
+    if(msg.value != valAccumulator) revert ValueMismatch();
+    emit Log(msg.sender, msg.value, "called batchSendEth");
+  }
+
+  function batchSendTokens(uint256 _tollCommitment, SendToken[] calldata _params) public payable {
+
+    for (uint256 i = 0; i < _params.length; i++) {
+      uint256 _amount = _params[i].amount;
+      address _tokenAddr = _params[i].tokenAddr;
+      IERC20 token = IERC20(address(_tokenAddr));
+
+      token.transferFrom(msg.sender, address(this), _amount);
+
+      if (token.allowance(address(this), address(umbra)) == 0) {
+        token.approve(address(umbra), type(uint256).max);
+      }
+
+      umbra.sendToken{value: _tollCommitment}(
+        _params[i].receiver,
+        _params[i].tokenAddr,
+        _amount,
+        _params[i].pkx,
+        _params[i].ciphertext
+      );
+    }
+    emit Log(msg.sender, msg.value, "called batchSendTokens");
+  }
+
+  function batchSend(
+    uint256 _tollCommitment,
+    SendEth[] calldata _ethParams,
+    SendToken[] calldata _tokenParams
+  ) external payable {
+    batchSendEth(_tollCommitment, _ethParams);
+    batchSendTokens(_tollCommitment, _tokenParams);
```

Comment:
Once you add the above `_tollCommitment * _params.length == msg.value` check, this method probably will not work due to conflicting `msg.value checks. Instead we'll need to refactor into the following interface:
- `_batchSendEth` is an internal method with no safety checks and executes all logic
- `batchSendETH` is an external method that executes safety checks the calls `_batchSendEth` 
- same pattern as above for `batchSendTokens`
- `batchSend` is an `external` method with its own safety checks, which then calls the two internal methods

This may result in us having to loop through arrays twice—once for safety checks, then a second time for execution—but I think that's ok for now and we can optimize in a future PR

Reply:
Looks like this was done, also still needs some tests on the safety checks. Also looks like `batchSendEth` and `batchSendToken` are `public`, but they can now be `external`